lightly/models/modules/memory_bank.py

```diff
@@ -48,38 +54,67 @@ class MemoryBankModule(torch.nn.Module):
 
     """
 
-    def __init__(self, size: int = 65536, gather_distributed: bool = False):
-        super(MemoryBankModule, self).__init__()
-
-        if size < 0:
-            msg = f"Illegal memory bank size {size}, must be non-negative."
-            raise ValueError(msg)
-
-        self.size = size
+    def __init__(
+        self,
+        size: Union[int, Sequence[int]] = 65536,
+        gather_distributed: bool = False,
+        feature_dim_first: bool = True,
+    ):
+        super().__init__()
+        size_tuple = (size,) if isinstance(size, int) else tuple(size)
+
+        if any(x < 0 for x in size_tuple):
+            raise ValueError(
+                f"Illegal memory bank size {size}, all entries must be non-negative."
+            )
+
+        self.size = size_tuple
         self.gather_distributed = gather_distributed
+        self.feature_dim_first = feature_dim_first
+        self.bank: Tensor
         self.register_buffer(
-            "bank", tensor=torch.empty(0, dtype=torch.float), persistent=False
+            "bank",
+            tensor=torch.empty(size=self.size, dtype=torch.float),
+            persistent=False,
         )
+        self.bank_ptr: Tensor
         self.register_buffer(
-            "bank_ptr", tensor=torch.empty(0, dtype=torch.long), persistent=False
+            "bank_ptr",
+            tensor=torch.empty(1, dtype=torch.long),
+            persistent=False,
         )
 
+        if isinstance(size, int) and size > 0:
+            warnings.warn(
+                (
+                    f"Memory bank size 'size={size}' does not specify feature "
+                    "dimension. It is recommended to set the feature dimension with "
+                    "'size=(n, dim)' when creating the memory bank. Distributed "
+                    "training might fail if the feature dimension is not set."
+                ),
+                UserWarning,
+            )
+        else:
+            self._init_memory_bank(dim=None)
+
     @torch.no_grad()
-    def _init_memory_bank(self, dim: int) -> None:
-        """Initialize the memory bank if it's empty
+    def _init_memory_bank(self, dim: Union[Sequence[int], None]) -> None:
```

Comment:
This is a bit convoluted I think it'd be a better solution to pass the `size` explicitly:
```python
    @torch.no_grad()
    def _init_memory_bank(self, size: Sequence[int]) -> None:
        """Initialize the memory bank if it's empty.
        Args:
            size:
                The dimensions of the memory bank.
        """
        self.bank = torch.randn(size).type_as(self.bank)
        self.bank = torch.nn.functional.normalize(self.bank, dim=-1)
        self.bank_ptr = torch.zeros(1).type_as(self.bank_ptr)
```

Also, we should initialize the memory bank when the size is passed in the constructor is a tuple.

Reply:
Updated :)

> Also, we should initialize the memory bank when the size is passed in the constructor is a tuple.

This is already implemented.